x/stake/validator.go

```diff
@@ -0,0 +1,242 @@
+)
+
+// Validator defines the total amount of bond shares and their exchange rate to
+// coins. Accumulation of interest is modelled as an in increase in the
+// exchange rate, and slashing as a decrease.  When coins are delegated to this
+// validator, the validator is credited with a Delegation whose number of
+// bond shares is based on the amount of coins delegated divided by the current
+// exchange rate. Voting power can be calculated as total bonds multiplied by
+// exchange rate.
+type Validator struct {
+	Status  sdk.BondStatus `json:"status"`  // bonded status
+	Address sdk.Address    `json:"address"` // sender of BondTx - UnbondTx returns here
+	PubKey  crypto.PubKey  `json:"pub_key"` // pubkey of validator
+
+	PShares         PoolShares `json:"pool_shares"`      // total shares for tokens held in the pool
+	DelegatorShares sdk.Rat    `json:"delegator_shares"` // total shares issued to a validator's delegators
+
+	Description        Description `json:"description"`            // description terms for the validator
+	BondHeight         int64       `json:"validator_bond_height"`  // earliest height as a bonded validator
+	BondIntraTxCounter int16       `json:"validator_bond_counter"` // block-local tx index of validator change
+	ProposerRewardPool sdk.Coins   `json:"proposer_reward_pool"`   // XXX reward pool collected from being the proposer
+
+	Commission            sdk.Rat `json:"commission"`              // XXX the commission rate of fees charged to any delegators
+	CommissionMax         sdk.Rat `json:"commission_max"`          // XXX maximum commission rate which this validator can ever charge
+	CommissionChangeRate  sdk.Rat `json:"commission_change_rate"`  // XXX maximum daily increase of the validator commission
+	CommissionChangeToday sdk.Rat `json:"commission_change_today"` // XXX commission rate change today, reset each day (UTC time)
+
+	// fee related
+	PrevBondedShares sdk.Rat `json:"prev_bonded_shares"` // total shares of a global hold pools
+}
+
+// Validators - list of Validators
+type Validators []Validator
+
+// NewValidator - initialize a new validator
+func NewValidator(address sdk.Address, pubKey crypto.PubKey, description Description) Validator {
+	return Validator{
+		Status:                sdk.Unbonded,
+		Address:               address,
+		PubKey:                pubKey,
+		PShares:               NewUnbondedShares(sdk.ZeroRat()),
+		DelegatorShares:       sdk.ZeroRat(),
+		Description:           description,
+		BondHeight:            int64(0),
+		BondIntraTxCounter:    int16(0),
+		ProposerRewardPool:    sdk.Coins{},
+		Commission:            sdk.ZeroRat(),
+		CommissionMax:         sdk.ZeroRat(),
+		CommissionChangeRate:  sdk.ZeroRat(),
+		CommissionChangeToday: sdk.ZeroRat(),
+		PrevBondedShares:      sdk.ZeroRat(),
+	}
+}
+
+// only the vitals - does not check bond height of IntraTxCounter
+func (v Validator) equal(c2 Validator) bool {
+	return v.Status == c2.Status &&
```

Comment:
can we just marshal for equality?

Reply:
Ah this function is checks everything besides bondheight/intraTxCounter - otherwise we would be able to

Reply:
implemented marshal for equality in some other locations with like functions